include/merlin/memory_pool.hpp

```diff
@@ -0,0 +1,600 @@
+ */
+template <class T>
+struct StandardAllocator final : AllocatorBase<T, StandardAllocator<T>> {
+  using type = typename AllocatorBase<T, StandardAllocator<T>>::type;
+
+  static constexpr const char* name{"StandardAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    return new type[n];
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    delete[] ptr;
+  }
+};
+
+/**
+ * Claim/release buffers in pinned host memory.
+ */
+template <class T>
+struct HostAllocator final : AllocatorBase<T, HostAllocator<T>> {
+  using type = typename AllocatorBase<T, HostAllocator<T>>::type;
+
+  static constexpr const char* name{"HostAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    CUDA_CHECK(cudaMallocHost(&ptr, sizeof(T) * n));
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    CUDA_CHECK(cudaFreeHost(ptr));
+  }
+};
+
+/**
+ * Claim/release buffers in the active CUDA device. Will not test if the correct
+ * device was used, and throw if CUDA runtime API response is negative.
+ */
+template <class T>
+struct DeviceAllocator final : AllocatorBase<T, DeviceAllocator<T>> {
+  using type = typename AllocatorBase<T, DeviceAllocator<T>>::type;
+
+  static constexpr const char* name{"DeviceAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    void* ptr;
+    cudaError_t res;
+    if (ctx) {
+      res = cudaMallocAsync(&ptr, sizeof(T) * n, ctx);
+    } else {
+      res = cudaMalloc(&ptr, sizeof(T) * n);
+    }
+    CUDA_CHECK(res);
+    return reinterpret_cast<type*>(ptr);
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    cudaError_t res;
+    if (ctx) {
+      res = cudaFreeAsync(ptr, ctx);
+    } else {
+      res = cudaFree(ptr);
+    }
+    CUDA_CHECK(res);
+  }
+};
+
+/**
+ * Wrapper around another allocator that prints debug messages.
+ */
+template <class Allocator>
+struct DebugAllocator final
+    : AllocatorBase<typename Allocator::type, DebugAllocator<Allocator>> {
+  using type = typename Allocator::type;
+
+  static constexpr const char* name{"DebugAllocator"};
+
+  inline static type* alloc(size_t n, cudaStream_t ctx = nullptr) {
+    type* ptr = Allocator::alloc(n, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " allocated = " << n
+              << " x " << sizeof(type) << " bytes, ctx = " << ctx << std::endl;
+    return ptr;
+  }
+
+  inline static void free(type* ptr, cudaStream_t ctx = nullptr) {
+    Allocator::free(ptr, ctx);
+    std::cout << Allocator::name << "[type_name = " << typeid(type).name()
+              << "]: " << static_cast<void*>(ptr) << " freed, ctx = " << ctx
+              << std::endl;
+  }
+};
+
+/**
+ * Helper structure to configure a memory pool.
+ */
+struct MemoryPoolOptions {
```

Comment:
For making the struct visible to the end-users and Doc generator, we should move the declaration to the `merlin_hashtable.cuh`

Reply:
Again, I am opposed to this. It would change the compilation order. Also, it would prevent compiling the resource pooling with GCC. Extend doc generator to render memory_pool.hpp would be better.

Reply:
See my comment above: `https://github.com/NVIDIA-Merlin/HierarchicalKV/pull/69#discussion_r1044154774`

Reply:
OK, I agree with you on compiling concern, but MemoryPoolOptions is an end-user interface, You'd better put them into a high-level header like `merlin-hashtable.cuh` . Or give a method to end-user to know how to use and set up the parameters.

Reply:
How about we add a `config.hpp` or `options.hpp` header where we collect all **config** or **options** structs?

Reply:
@rhdong ?